### PR TITLE
Created Board-Logic Wrapper (issue#37)

### PR DESCRIPTION
**Pull request by TrevorPeters**
created core/game/board.go and core/game/board_test.go.

### CODE REVIEW

File: core/game/board.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
+package game
```
Comment (artasparks) [inline core/game/board.go]: Cane we move this to core/board.go?

File: core/game/board.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+	"github.com/otrego/clamshell/core/point"
+)
+
+// Board Contains the board, capturesStones, and ko
+// capturedStones only contains the stones captured
+// from the most recent AddStone call. ko contains
+// a point that is illegal to recapture due to Ko.
+type Board struct {
+	board          [][]color.Color
+	capturedStones []*point.Point
```
Comment (artasparks) [inline core/game/board.go]: Don't store captured stones here -- they should be passed as method params.

File: core/game/board.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+	stoneGroup := make([]*point.Point, len(expanded))
+	i := 0
+	for key := range expanded {
+		stoneGroup[i] = point.New(key.X(), key.Y())
+		i++
+	}
+	return stoneGroup
+}
+
+// String returns a string representation of this board.
```
Comment (artasparks) [inline core/game/board.go]: Can you add an example of what this looks like?

File: core/game/board.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+func (b *Board) AddStone(m *Move) ([]*point.Point, error) {
+	pt := m.Point()
+	var (
+		x, y int          = int(pt.X()), int(pt.Y())
+		ko   *point.Point = b.ko
+	)
+	b.ko = nil
+	b.capturedStones = nil
+
+	if x >= len(b.board[0]) || y >= len(b.board) ||
```
Comment (artasparks) [inline core/game/board.go]: I think an inBounds or outOfBounds helper would be useful here.

File: core/game/board.go [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+		x < 0 || y < 0 {
+		return nil, fmt.Errorf("move %v out of bounds for %dx%d board",
+			pt, len(b.board[0]), len(b.board))
+	}
+	if b.board[y][x] != color.Empty {
```
Comment (artasparks) [inline core/game/board.go]: how about having a b.colorAt(pt) helper?

File: core/game/board.go [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+		return nil, fmt.Errorf("move %v already occupied", pt)
+	}
+
+	b.board[y][x] = m.Color()
+	b.FindCapturedGroups(m)
```
Comment (artasparks) [inline core/game/board.go]: this should return the captured groups rather than dumping into a field in the struct.

File: core/game/board.go [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+	b.RemoveCapturedStones()
+	return b.capturedStones, nil
+}
+
+// FindCapturedGroups finds the groups captured by *Move m.
+func (b *Board) FindCapturedGroups(m *Move) {
+	pt := m.Point()
+
+	points := make([]*point.Point, 4)
+	points[0] = point.New(pt.X()+1, pt.Y())
```
Comment (artasparks) [inline core/game/board.go]: I would recommend a neighbors method `getNeighbors(pt)` => []*point.Point

File: core/game/board.go [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+			return nil, fmt.Errorf("%v is an illegal ko move", pt)
+		}
+	}
+
+	b.RemoveCapturedStones()
+	return b.capturedStones, nil
+}
+
+// FindCapturedGroups finds the groups captured by *Move m.
+func (b *Board) FindCapturedGroups(m *Move) {
```
Comment (artasparks) [inline core/game/board.go]: should be private

File: core/game/board.go [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+	capturedStones := make([]*point.Point, 0)
+	for _, point := range points {
+		capturedStones = append(capturedStones, b.CapturedStones(point)...)
+	}
+	b.capturedStones = capturedStones
+}
+
+// RemoveCapturedStones removes the captured stones from
+// the board.
+func (b *Board) RemoveCapturedStones() {
```
Comment (artasparks) [inline core/game/board.go]: should be private and take the stones as a parameter

File: core/game/board.go [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+	for _, point := range b.capturedStones {
+		b.board[point.Y()][point.X()] = color.Empty
+	}
+}
+
+// CapturedStones returns the captured stones in group containing Point pt.
+// returns nil if no stones were captured
+func (b *Board) CapturedStones(pt *point.Point) []*point.Point {
```
Comment (artasparks) [inline core/game/board.go]: private

**Review by artasparks — COMMENTED** [review]
Some initial comments

File: core/game/board.go (hunk 1 above)
Comment (TrevorPeters) [reply core/game/board.go]: Just to be sure, you mean move this file from core/game to core/board?

File: core/board/board.go [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+}
+
+// AddStone adds a stone to the board.
+// returns the captured stones, or err
+// if any Go (baduk) rules were broken
+func (b *Board) AddStone(m *game.Move) ([]*point.Point, error) {
+	var ko *point.Point = b.ko
+	b.ko = nil
+
+	if !b.inBounds(m.Point()) {
```
Comment (artasparks) [inline core/board/board.go:40]: nice

File: core/board/board.go [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+		pt1, ok := e.Value.(*point.Point)
+		if !ok {
+			panic("e.Value was not of type point.Point")
+		}
+
+		if !b.inBounds(pt1) {
+			continue
+		} else if b.colorAt(pt1) == color.Empty {
+			// Liberty has been found, no need to continue search
+			return nil
```
Comment (artasparks) [inline core/board/board.go:108]: Nice.

**Review by artasparks — APPROVED** [review]
This is rad, nicely done.

**Pull request merged**